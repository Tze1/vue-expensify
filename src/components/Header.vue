<template>
  <header class="header component">
    <div class="masthead"><h1>{{ appTitle ? appTitle : 'VueJS App' }}</h1></div>
    <div v-show="userLoggedIn" class="nav">
      <router-link to="/dashboard">Dashboard</router-link>|
      <router-link to="/add">Add Expense</router-link>
    </div>
    <div v-if="user" class="auth">
      <span class="username">{{ user ? user.displayName : '' }}</span>
      <img :src="getAvatarUrl()" class="avatar"/>
      <button
        @click="startLogout"
        aria-label="Log Out"
        title="Log Out"
        class="logout-btn"
      >
        <font-awesome-icon icon="sign-out-alt" />
      </button>
    </div>
    <DismissableAlert />
  </header>
</template>

<script>
  import { mapState } from 'vuex';
  import DismissableAlert from './DismissableAlert';

  export default {
    name: 'Header',
    props: {
      appTitle: {
        type: String,
        required: true,
      },
      startLogout: {
        type: Function,
        required: true,
      },
    },
    computed: mapState({
      userLoggedIn () {
        return this.$store.getters.userLoggedIn;
      },
      user: state => state.auth.user,
    }),
    methods: {
      getAvatarUrl () {
        return this.user.photoURL ?
          this.user.photoURL :
          '/img/avatar-default.png';
      },
    },
    components: {
      DismissableAlert,
    },
  };
</script>

<style scoped lang="scss">
  @import '../styles/app.scss';

  .header.component {
    align-items: flex-start;
    display: flex;
    flex-flow: column nowrap;
    justify-content: flex-start;
    padding: $space-lg 0 $space-md;
    text-align: left;

    @include mq("tablet") {
      align-items: baseline;
      flex-flow: row nowrap;
    }

    > div {
      flex: 0 0 auto;
      padding: 0 $space-xs $space-xs 0;
      vertical-align: baseline;

      &:first-child {
        padding-left: 0;
        padding-top: 0;
      }

      &.auth {
        align-self: flex-end;
        padding-bottom: 0;
        padding-right: 0;
        text-align: right;
      }

      &.dismissablealert.component {
        padding: 0.75rem 1.25rem;
      }

      @include mq("tablet") {
        padding: 0 $space-sm;

        &.auth {
          align-self: initial;
        }
      }
    }  // > div

    .masthead {
      h1 {
        background: transparent url('/img/logo.png') left 50% no-repeat;
        background-size: 1.75rem 1.75rem;
        padding-left: 2.5rem;
        line-height: 1;
        margin: 0;
      }
    }

    .nav {
      @include mq("tablet") {
        flex-grow: 1;
      }

      a {
        margin: 0 $space-xs;

        &:first-child {
          margin-left: 0;
        }

        &:last-child {
          margin-right: 0;
        }

        &.router-link-active {
          font-weight: bold;
          text-decoration: none;
        }
      }
    }

    .auth {
      text-align: right;
      margin-top: -3.8rem;
      width: 100%;

      @include mq("tablet") {
        margin-top: 0;
        width: auto;
      }

      > * {
        margin-left: $space-xs;

        @include mq("tablet") {
          margin-left: $space-sm;
        }

        &:first-child {
          margin-left: 0;
        }
      }

      .username {
        display: none;
        font-weight: bold;

        @include mq("tablet") {
          display: inline-block;
        }
      }

      .avatar {
        border-radius: 50%;
        height: 2rem;
        vertical-align: middle;
        width: 2rem;
      }

      .logout-btn {
        background: none;
        border: none;
        font-size: $font-size-xl;
        line-height: $font-size-xl;
        margin-right: 0;
        padding: 0;

        > .svg-inline--fa {
          color: $gray4;
          vertical-align: middle;

          &:hover {
            color: $white;
          }
        }
      }
    } // .auth
  }  // header
</style>
