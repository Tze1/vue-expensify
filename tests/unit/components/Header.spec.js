import { shallowMount } from '@vue/test-utils'
import Header from '@/components/Header.vue'

describe('Header.vue', () => {
  const $store = {
    state: {
      auth: {
        user: {
          displayName: 'Test User',
          photoURL: '/img/avatar-default.png',
        },
      },
    },
    getters: {
      userLoggedIn: state => {
        return (state.auth.user !== undefined);
      },
    }
  }
  const appTitle = 'Test-App'
  const startLogoutStub = jest.fn();
  const wrapper = shallowMount(Header, {
    mocks: {
      $store,
    },
    stubs: [
      'router-link',
      'font-awesome-icon',
    ],
    propsData: {
      appTitle,
      startLogout: startLogoutStub,
    },
  })

  it('renders Header component properly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('calls startLogout properly when logout-btn is clicked', () => {
    wrapper.find('.logout-btn').trigger('click');

    expect(startLogoutStub).toHaveBeenCalled();
  });
})
