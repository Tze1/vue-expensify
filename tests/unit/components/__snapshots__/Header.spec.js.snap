// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header.vue renders Header component properly 1`] = `
<header class="header component">
  <div class="masthead">
    <h1>Test-App</h1>
  </div>
  <div class="nav">
    <router-link-stub to="/dashboard">Dashboard</router-link-stub> |
    <router-link-stub to="/add">Add Expense</router-link-stub>
  </div>
  <div class="auth"><span class="username">Test User</span> <img src="/img/avatar-default.png" class="avatar"> <button class="button logout-btn exit">Logout</button></div>
</header>
`;
